native/c/zam.c

```diff
@@ -30,6 +38,830 @@ RDJFCB_MODEL(rdfjfcb_model);
+    diag->detail_rc = ZDS_RTNCD_RESERVE_ERROR;
+    return RTNCD_FAILURE;
+  }
+
+  ioc->reserve = 1;
+  return rc;
+}
+
+static int open_data_set(ZDIAG *PTR32 diag, IO_CTRL *PTR32 ioc)
+{
+  int rc = 0;
+  rc = open_output(&ioc->dcb);
+  if (0 != rc)
+  {
+    diag->service_rc = rc;
+    strcpy(diag->service_name, "OPEN");
+    diag->e_msg_len = sprintf(diag->e_msg, "Failed to open ddname: %8.8s for data set: %44.44s rc was: %d", ioc->dcb.dcbddnam, ioc->jfcb.jfcbdsnm, rc);
+    diag->detail_rc = ZDS_RTNCD_OPEN_ERROR;
+    return RTNCD_FAILURE;
+  }
+
+  if (!(ioc->dcb.dcboflgs & dcbofopn))
+  {
+    diag->e_msg_len = sprintf(diag->e_msg, "Data set is not open: %44.44s", ioc->jfcb.jfcbdsnm);
+    diag->detail_rc = ZDS_RTNCD_NOT_OPEN_ERROR;
+    return RTNCD_FAILURE;
+  }
+
+  return rc;
+}
+
+static int validate_dcb_attributes(ZDIAG *PTR32 diag, IO_CTRL *PTR32 ioc)
+{
+  int rc = 0;
+
+  if (!(ioc->dcb.dcbrecfm & (dcbrecf | dcbrecv)))
+  {
+    diag->e_msg_len = sprintf(diag->e_msg, "Data set is not a fixed or variable record format: %X", ioc->dcb.dcbrecfm);
+    diag->detail_rc = ZDS_RTNCD_UNSUPPORTED_RECFM;
+    return RTNCD_FAILURE;
+  }
+
+  int block_size = ioc->dcb.dcbblksi;
+
+  if (block_size < 1)
+  {
+    diag->e_msg_len = sprintf(diag->e_msg, "Data set has less than 1 block size: %X", block_size);
+    diag->detail_rc = ZDS_RTNCD_UNSUPPORTED_BLOCK_SIZE;
+    return RTNCD_FAILURE;
+  }
+
+  if (block_size % ioc->dcb.dcblrecl != 0)
+  {
+    // if the data set is not blocked, we can subtract the size of the RDW
+    if (!(ioc->dcb.dcbrecfm & dcbrecbr))
+    {
+      block_size -= sizeof(BDW);
+
+      // perform the check again
+      if (block_size % ioc->dcb.dcblrecl != 0)
+      {
+        diag->e_msg_len = sprintf(diag->e_msg, "Data set block size (minus 4) is not a multiple of the record length: %d not evenly divisible by %d", ioc->dcb.dcbblksi, ioc->dcb.dcblrecl);
+        diag->detail_rc = ZDS_RTNCD_INVALID_BLOCK_SIZE;
+        return RTNCD_FAILURE;
+      }
+    }
+    else
+    {
+      diag->e_msg_len = sprintf(diag->e_msg, "Data set block size is not a multiple of the record length: %d not evenly divisible by %d", ioc->dcb.dcbblksi, ioc->dcb.dcblrecl);
+      diag->detail_rc = ZDS_RTNCD_INVALID_BLOCK_SIZE;
+      return RTNCD_FAILURE;
+    }
+  }
+
+  return rc;
+}
+
+static int note_member(ZDIAG *PTR32 diag, IO_CTRL *PTR32 ioc, NOTE_RESPONSE *PTR32 note_response)
+{
+
+  int rc = 0;
+  int rsn = 0;
+
+  rc = note(ioc, note_response, &rsn);
+  if (0 != rc)
+  {
+    if (0 == diag->e_msg_len)
+    {
+      diag->service_rc = rc;
+      strcpy(diag->service_name, "NOTE");
+      diag->e_msg_len = sprintf(diag->e_msg, "Failed to NOTE ddname: %8.8s data set: %44.44s rc was: %d", ioc->ddname, ioc->jfcb.jfcbdsnm, rc);
+      diag->detail_rc = ZDS_RTNCD_NOTE_ERROR;
+    }
+  }
+
+  return rc;
+}
+
+int open_output_bpam(ZDIAG *PTR32 diag, IO_CTRL *PTR32 *PTR32 ioc, const char *PTR32 ddname)
```

Comment:
A common pattern with file handles in the C and C++ standard library is that if the return code is non-zero for a function like `fopen`, the caller doesn't need to call `fclose`. Would it be possible to establish a similar pattern here so that `close_output_bpam` is called within this function if a blocking error was encountered (cannot continue using BPAM)? The need to call `close_output_bpam` wasn't obvious to me when first reading the `zds` code but its becoming clearer as I peruse through the changed files.

I think it makes sense for the user to have to call `close_...` once they are done, but having to call `close` after an error in `open` seems to deviate from the aforementioned std library convention.

Reply:
Nice, I think the metal C API is a great place to put it 👍 thanks!